### PR TITLE
[Fixes #285] Allow calling pydoctor multiple times in Sphinx extension.

### LINKED ISSUE #285: Output samples
@adiroiban [wrote](https://github.com/twisted/pydoctor/pull/282#pullrequestreview-537405656):
> My suggestion is to create a pydoctor.test.sample.epydoc or pydoctor.samples.epydoc package in which we include all the possible usage of pydoctor with epydoc
>
> In this way, the result of this PR could be observed via
> https://pydoctor--282.org.readthedocs.build/en/282/api/pydoctor.test.html

@mthuurne [wrote](https://github.com/twisted/pydoctor/pull/282#issuecomment-732943736):
> I think it would be good to have a package that displays all the supported use cases, but I think it should be part of the documentation rather than part of pydoctor's code.
>
> One reason is that we'd want to showcase syntax that is not available in all Python versions, such as position-only parameters. Having that syntax in the main source dir could cause problems with distros that pre-optimize Python modules on install.
>
> Another reason is that I wouldn't want samples to be mixed into the indexes of the actual API docs. In the package index it would be just one line, but in the class hierarchy and name index it would produce more noise.

### PR DESCRIPTION
**Pull request by adiroiban**
This tries to fix #285 

I expect that 99% of pydoctor users will want to call pydoctor a single time.
This is why for those 99% cases, the interface is just a simple list.

for the other 1% you can configure the extension with a dict.

Please suggest better names for package / directories.

For now, it only provides a limited sample for epydoc/epytext.
We can extend the sample as we work on various PRs.

Support for RST can be added  in a separate PR. I am not interested in working on RST :)

Result at https://pydoctor--314.org.readthedocs.build/en/314/
https://pydoctor--314.org.readthedocs.build/en/314/docformat/epytext/demo/index.html

### CODE REVIEW

**Comment by tristanlatr** [comment]
Little remark: I would not put the links to the demo in the quick links TOC, I think it's not that primordial. We could have the links in the respective docformat pages perhaps? At the beginning.

**Comment by adiroiban** [comment]
I have reduced the scope of this PR ....

We now only have https://pydoctor--314.org.readthedocs.build/en/314/docformat/epytext/demo/index.html

This is ready for review.

**Review by mthuurne — APPROVED** [review]
It would be useful to have source links in the generated docs. I'm not sure if we can easily do relative source links though and predicting absolute URLs might complicate the build. I guess in the worst case you could pass a dummy base URL though and then replace it in the output with `sed`.

Regarding the `pydoctor_args` config item: in general I'm not a big fan of values that can have more than one type. But in this case, it is indeed likely that the vast majority of users will only run `pydoctor` once and requiring the more complex form from them would not be user-friendly. So I think having two types is fine here.

I'm approving this because I think it doesn't need a second review round. The `dict` to `Mapping` change should be done before merging though, in my opinion. The formatting and spelling fixes would be nice to include but could also be done later as they don't impact the interface.

File: docs/epytext_demo/__init__.py
Comment (adiroiban) [reply docs/epytext_demo/__init__.py:86]: This is on purpose... as a test that we really have literal blocks... and that the X tag is not interpreted.

Comment (mthuurne) [reply docs/epytext_demo/__init__.py:86]: No, it's fine in that case. Reviewing epytext that is also documenting epytext is just confusing ;)

File: docs/epytext_demo/__init__.py
Comment (adiroiban) [reply docs/epytext_demo/__init__.py:19]: I have updated... hope I fixed all this type of errors.

File: docs/epytext_demo/__init__.py
Comment (adiroiban) [reply docs/epytext_demo/__init__.py:96]: hope this is fixed :) thanks

File: docs/epytext_demo/__init__.py
Comment (adiroiban) [reply docs/epytext_demo/__init__.py]: I don't know :)

File: docs/epytext_demo/demo_epytext_module.py
Comment (adiroiban) [reply docs/epytext_demo/demo_epytext_module.py:24]: I tried to update the examples. Feel free to make any changes here.

File: docs/epytext_demo/demo_epytext_module.py
Comment (adiroiban) [reply docs/epytext_demo/demo_epytext_module.py:38]: I will give it a try.

File: docs/epytext_demo/demo_epytext_module.py
Comment (adiroiban) [reply docs/epytext_demo/demo_epytext_module.py:41]: Yep. Thanks

File: pydoctor/sphinx_ext/build_apidocs.py
Comment (adiroiban) [reply pydoctor/sphinx_ext/build_apidocs.py:10]: I would like to keep  the dict usage ... as semi-private. Most project should not need it.
Maybe not even documenting this at all here.

File: docs/epytext_demo/__init__.py
Comment (mthuurne) [inline docs/epytext_demo/__init__.py:20]:
```suggestion
General epytext formating markups are documented in the docstring of this package.
```

File: docs/epytext_demo/__init__.py
Comment (mthuurne) [inline docs/epytext_demo/__init__.py:22]:
```suggestion
Epydoc code related formating are demonstrated in the modules of this package.
```

File: pydoctor/sphinx_ext/build_apidocs.py
Comment (mthuurne) [inline pydoctor/sphinx_ext/build_apidocs.py:56]: Do you need to modify the arguments? If not, it's better to use an immutable type such as `Sequence` or `Iterable`.

**Review by mthuurne — COMMENTED** [review]
I found a few remaining typos and one case where an annotation could maybe be made more abstract.

**Pull request merged**
